app/assets/javascripts/discourse/app/mixins/composer-video-thumbnail-uppy.js

```diff
@@ -0,0 +1,122 @@
+  _generateVideoThumbnail() {
+    let videos = document.getElementsByClassName("video-container");
+    if (!videos) {
+      return;
+    }
+
+    let video_container = videos[0];
+    if (!video_container) {
+      return;
+    }
+
+    let video = video_container.querySelector("video:first-of-type");
+    if (!video) {
+      return;
+    }
+
+    let video_src = video.getElementsByTagName("source")[0].src;
+    let video_sha1 = video_src
+      .substring(video_src.lastIndexOf("/") + 1)
+      .split(".")[0];
+
+    // Wait for the video element to load, otherwise the canvas will be empty
+    video.oncanplay = () => {
+      let canvas = document.createElement("canvas");
+      let ctx = canvas.getContext("2d");
+      let videoHeight, videoWidth;
+      videoHeight = video.videoHeight;
+      videoWidth = video.videoWidth;
+      canvas.width = videoWidth;
+      canvas.height = videoHeight;
+
+      ctx.drawImage(video, 0, 0, videoWidth, videoHeight);
+
+      // upload video thumbnail
+      canvas.toBlob((blob) => {
+        this._uppyInstance = new Uppy({
+          id: `screenshot-placeholder`,
+          meta: {
+            upload_type: `thumbnail`,
+            video_sha1,
+          },
+          autoProceed: true,
+        });
+
+        if (this.siteSettings.enable_upload_debug_mode) {
+          this._instrumentUploadTimings();
+        }
+
+        if (this.siteSettings.enable_direct_s3_uploads) {
+          this._useS3MultipartUploads();
+        } else {
+          this._useXHRUploads();
+        }
+        this._uppyInstance.use(DropTarget, { target: this.element });
+
+        this._uppyInstance.on("upload", () => {
+          this.set("uploading", true);
+        });
+
+        this._uppyInstance.on("upload-success", () => {
+          this.set("uploading", false);
+        });
+
+        this._uppyInstance.on("upload-error", (file, error, response) => {
+          let message = I18n.t("wizard.upload_error");
+          if (response.body.errors) {
+            message = response.body.errors.join("\n");
+          }
+
+          this.dialog.alert(message);
+          this.set("uploading", false);
+        });
+
+        try {
+          this._uppyInstance.addFile({
+            source: `${this.id} thumbnail`,
+            name: video_sha1,
+            type: blob.type,
+            data: blob,
+          });
+        } catch (err) {
+          warn(`error adding files to uppy: ${err}`, {
+            id: "discourse.upload.uppy-add-files-error",
+          });
+        }
+      });
+    };
+  },
+
+  // This should be overridden in a child component if you need to
+  // hook into uppy events and be sure that everything is already
+  // set up for _uppyInstance.
+  _uppyReady() {},
+
+  _useXHRUploads() {
+    this._uppyInstance.use(XHRUpload, {
+      endpoint: getURL(`/uploads.json?client_id=${this.messageBus.clientId}`),
+      headers: {
+        "X-CSRF-Token": this.session.csrfToken,
+      },
+    });
+  },
+});
```

Comment:
Is it possible to test this somewhat? There is a bit of frontend/backend interaction here which makes it tricky, but a useful test would be to have a frontend test of a video file in the composer that triggers an uppy upload. 

Or, possibly a system test, the tricky thing there will be to find a way to display the thumbnail somewhere, I'm not sure we can easily do that in core (without theme components or plugins). 

(That said, given this feature is experimental, I'm OK with handling the test in a followup PR.)

Reply:
> (That said, given this feature is experimental, I'm OK with handling the test in a followup PR.)

:+1: I'm going to hold off for now then. I definitely want an end-to-end test, just not sure it's worth the effort at this stage.